synapse/storage/databases/main/account_data.py

```diff
@@ -424,6 +421,70 @@ async def add_account_data_for_user(
 
         return self._account_data_id_gen.get_current_token()
 
+    def _add_account_data_for_user(
+        self,
+        txn,
+        next_id: int,
+        user_id: str,
+        account_data_type: str,
+        content: JsonDict,
+    ) -> None:
+        content_json = json_encoder.encode(content)
+
+        # no need to lock here as account_data has a unique constraint on
+        # (user_id, account_data_type) so simple_upsert will retry if
+        # there is a conflict.
+        self.db_pool.simple_upsert_txn(
+            txn,
+            table="account_data",
+            keyvalues={"user_id": user_id, "account_data_type": account_data_type},
+            values={"stream_id": next_id, "content": content_json},
+            lock=False,
+        )
+
+        # Ignored users get denormalized into a separate table as an optimisation.
+        if account_data_type == AccountDataTypes.IGNORED_USER_LIST:
+            # Insert / delete to sync the list of ignored users.
+            previously_ignored_users = set(
+                self.db_pool.simple_select_onecol_txn(
+                    txn,
+                    table="ignored_users",
+                    keyvalues={"ignorer_user_id": user_id},
+                    retcol="ignored_user_id",
+                )
+            )
+
+            # If the data is invalid, no one is ignored.
+            ignored_users_content = content.get("ignored_users", {})
+            if isinstance(ignored_users_content, dict):
+                currently_ignored_users = set(ignored_users_content)
+            else:
+                currently_ignored_users = set()
+            newly_ignored_users = currently_ignored_users - previously_ignored_users
+
+            # Delete entries which are no longer ignored.
+            self.db_pool.simple_delete_many_txn(
+                txn,
+                table="ignored_users",
+                column="ignored_user_id",
+                iterable=previously_ignored_users - currently_ignored_users,
+                keyvalues={"ignorer_user_id": user_id},
+            )
+
+            # Add entries which are newly ignored.
+            self.db_pool.simple_insert_many_txn(
+                txn,
+                table="ignored_users",
+                values=[
+                    {"ignorer_user_id": user_id, "ignored_user_id": u}
+                    for u in newly_ignored_users
+                ],
+            )
+
+            # Invalidate the cache for newly ignored users.
+            for ignored_user_id in newly_ignored_users:
+                self.ignored_by.invalidate((ignored_user_id,))
```

Comment:
This needs to be streamed to workers, so 
```suggestion
                self._invalidate_cache_and_stream(txn, self.ignored_by, (ignored_user_id,))
```

Reply:
I also realize this needs to be done for all changes (both added and removed users).